tests/unit_tests/test_controller_utils.py

```diff
@@ -65,6 +65,20 @@ def get_custom_controller_resources(keys, default):
             controller_resources_config, k, v)
 
 
+def _check_controller_resources(controller_resources, expected_combinations,
+                                default_controller_resources):
```

Comment:
Lets add type annotation for the arguments and return value?

Reply:
This is awesome! Can we change other unittest to use this function as well? (case 1,2,3)